sound/soc/sof/topology.c

```diff
@@ -492,6 +502,17 @@ static int get_token_u16(void *elem, void *object, u32 offset, u32 size)
 	return 0;
 }
 
+static int get_token_uuid(void *elem, void *object, u32 offset, u32 size)
+{
+	struct snd_soc_tplg_vendor_uuid_elem *velem = elem;
+	u32 *val = (u32 *)((u8 *)object + offset);
+	guid_t uuid;
+
+	guid_copy(&uuid, (guid_t *)velem->uuid);
```

Comment:
@andy-shev suggests in Skylake patches the use of import_guid(&uuid, velem->uuid);

Reply:
Yes, better to avoid explicit castings where it's not needed. And strictly speaking UUID types can hold more than simple buffer.

Reply:
Also `uuid_t` and `guid_t` shouldn't be present in any ABI headers (firmware interface is ABI).

Reply:
makes sense, thanks @plbossart @andy-shev

Reply:
The kernel is not using the UUID data other than passing it to the FW. So we need to use the same type/format used by the FW internally.
We will be keeping the subtype/flavor for the kernel to use for DAPM reasons (and this will also be in IPC2.0) e.g
```
/* kernel only cares about any component that changes DAPM routing */
enum sof_comp_type {
    SOF_TYPE_MUX,
    SOF_TYPE_MIXER,
    SOF_TYPE_SWITCH, 
    /* others ?? */
    SOF_TYPE_PROCESS, /* everything else goes here */
};
```
IPC2.0 will get rid of the need for sof_widget_load_type() from topology.c hence the reduced type/flavours above.

Reply:
@lgirdwood not quite get your point about IPC2.0, do you have a document about that?

Can you help to point which step in the documentation we should change here:
https://github.com/thesofproject/sof-docs/pull/245

As you know, it is 16 Bytes array from the UUID token, so if needed, we need to do a byte array -> struct sof_uuid transformation in the driver, why don't we leave this to firmware then it is totally transparent to the dirver?

```
/* vendor tuple for uuid */
struct snd_soc_tplg_vendor_uuid_elem {
	__le32 token;
	char uuid[16];
} __attribute__((packed));
```
```
struct sof_uuid {
	uint32_t a;
	uint16_t b;
	uint16_t c;
	uint8_t  d[8];
};
```

Reply:
```
struct sof_uuid {
	uint32_t a;
	uint16_t b;
	uint16_t c;
	uint8_t  d[8];
};
```
https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding
> The binary encoding of UUIDs varies between systems. Variant 1 UUIDs, nowadays the most common variant, are encoded in a big-endian format. For example, 00112233-4455-6677-8899-aabbccddeeff is encoded as the bytes 00 11 22 33 44 55 66 77 88 99 aa bb cc dd ee ff.
> Variant 2 UUIDs, historically used in Microsoft's COM/OLE libraries, use a mixed-endian format, whereby the first three components of the UUID are little-endian, and the last two are big-endian. [ROFL]

Reply:
@marc-hb I think we have aligned to use the 16 bytes array directly from UUID token, and leave the member designating and comparing to the Firmware internal.

Reply:
the uuid array will be in binary bytes so no endian issue here, mark it resolved.